DBZ-2897 Extended RecordCommitter interface to handle updated source offsets.

twthorn opened the pull request:
https://issues.redhat.com/browse/DBZ-2897

Extended the DebeziumEngine interface to include a new method for marking a record processed with updated source offsets. Implemented the new interface method in both ConvertingEngineBuilder and EmbeddedEngine. EmbeddedEngine creates a new SourceRecord with the updated sourceOffsets and passes this to the original markProcessed method. Added a unit test in EmbeddedEngineTest that verifies that the file offset storage contains the updated source offset’s partition. Added a revapi ignore case to prevent failure from a method added to an interface without a default method. Updated docs for Debezium Engine that describe the new functionality.

jpechane commented:
@gunnarmorling LGTM. The only question I have is whether the sources should be encapsulated into an another datatype effectivelly decoupling from the map but I am not really sure how this could be done.

gunnarmorling commented on debezium-embedded/src/test/java/io/debezium/embedded/EmbeddedEngineTest.java:
```diff
@@ -328,6 +331,80 @@ public void shouldRunDebeziumEngine() throws Exception {
+                }
+            }
+        }
+        assert containsCustomPartition;
```
Should use `assertThat()`.

twthorn replied:
Good point, thank you, updated.

gunnarmorling reviewed (commented):
One comment inline.

Re the offset representation as map, yeah, good point. We could have a simple contract like this instead:

```
interface Offsets {
    Offsets set(String key, Object value);
}
```

And pass in that instead of the map.

twthorn commented:
Updated to use assertThat rather than java assert keyword.

Should I also add the `Offsets` contract in this pull request? If so, could you provide some more info on where you’d want it defined, and where it would be implemented (e.g., whether by the client or by `EmbeddedEngine`)? Would this follow a similar pattern to `RecordChangeEvent` and `EmbeddedEngineChangeEvent`? Any additional info would be helpful.

gunnarmorling commented:
> Should I also add the Offsets contract in this pull request?

Yes, please.

> some more info on where you’d want it defined

I'd suggest an inner interface in `DebeziumEngine`, like `RecordCommitter` itself. The implementation would live next to where that one is implemented, too.

twthorn commented:
Added `Offsets` contract in `DebeziumEngine` and implemented it as `SourceRecordOffsets` in `EmbeddedEngine`. Added additional method `buildOffsets` in `RecordCommitter` for retrieving an `Offsets` object, implemented in `ConvertingEngineBuilder` and `EmbeddedEngine`. Modified `markProcessed` method to use `Offsets` contract. Updated tests and documentation.

gunnarmorling commented on debezium-embedded/src/test/java/io/debezium/embedded/EmbeddedEngineTest.java:
```diff
@@ -328,6 +330,81 @@ public void shouldRunDebeziumEngine() throws Exception {
+        try (BufferedReader br = new BufferedReader(new FileReader(OFFSET_STORE_PATH.toString()))) {
+            String line;
+            while ((line = br.readLine()) != null) {
+                logger.info(line);
```
So the line that's read back here is some binary representation, not plain text string. We should be able to get back the proper offset by reading this as a byte array and feeding it through `JsonConverter#toConnectData()` (ie. inverting the offset serialization step). While the assertion below technically passes, it feels a bit weird, and you could also do a proper lookup in the offset map then.

twthorn replied:
Good point, updated.

gunnarmorling commented on debezium-api/src/main/java/io/debezium/engine/DebeziumEngine.java:
```diff
@@ -121,6 +121,36 @@ default void taskStopped() {
+     * Contract that should be passed to {@link RecordCommitter#markProcessed(Object, Offsets)} for marking a record
+     * as processed with updated offsets.
+     */
+    public static interface Offsets {
```
```suggestion
    public interface Offsets {
```

It doesn't make a difference here.

twthorn replied:
Updated.

gunnarmorling reviewed (commented):
Thanks a lot, @twthorn. Looks good to me implementation-wise now. One small nit re the test assertion only; other than that, it's good to go.

twthorn commented:
Made the offsets interface non-static. Refactored the unit test for updating offsets to instead read the offset file as a map of byte array to byte array, parse the byte array as a json string, deserialize the json string, and make assertions on the key/values in the deserialized JSON object. Also minor refactors for some test constants to make assertions clearer.

gunnarmorling commented:
LGTM, thx! That test failure is unrelated and already tracked. Merging.
